theano/gof/graph.py

```diff
@@ -816,6 +816,92 @@ def clone_get_equiv(inputs, outputs, copy_inputs_and_orphans=True, memo=None):
     return memo
 
 
+def map_variables(fn, graphs, additional_inputs=[]):
+    """
+    Construct new graphs based on 'graphs' with some variables replaced
+    according to 'fn'.
+
+    :param fn: function that takes a variable and returns its replacement
+    :param graphs: an iterable of graphs in which to replace variables
+    :param additional_inputs: an iterable of graph inputs not used in any
+         of 'graphs' but possibly used in the graphs returned by 'fn'
+    :return: the new graphs, in the same order as 'graphs'
+
+    Example:
+
+    .. code-block:: python
+
+        import theano.tensor
+
+        tag = "replaceme"
+
+        a = theano.tensor.scalar("a")
+        b = theano.tensor.scalar("b")
+        c = theano.tensor.scalar("c")
+
+        ab = a + b
+        setattr(ab.tag, tag, True)
+
+        u = ab + c
+        v, = map_variables(
+            lambda x: a * b if getattr(x.tag, tag, False) else x,
+            [u])
+
+        # v is now equal to a * b + c
+    """
+    from fg import FunctionGraph
+    from opt import TopoOptimizer, local_optimizer
+    from theano import clone as the_other_clone
+    from theano.scan_module.scan_op import Scan
+
+    graphs = list(graphs)
+    inputs_ = list(set(inputs(graphs) + list(additional_inputs)))
+
+    # work on a copy of the graph, but ensure it is still in terms of
+    # the user's inputs.  we can use clone with copy_inputs=False to
+    # achieve this, but first we should clone those inputs that are
+    # cached constants, or FunctionGraph will complain about them.
+    cached_constants = [x for x in inputs_ if getattr(x, "cached", False)]
+    copied_constants, _ = clone(cached_constants, [], copy_inputs=True)
+    graphs = the_other_clone(graphs,
+                             share_inputs=True,
+                             replace=zip(cached_constants, copied_constants))
+    inputs_ = list(set(inputs_) - set(cached_constants)) + list(copied_constants)
+    fg = FunctionGraph(inputs_, graphs, clone=False)
+
+    nodes_seen = set()
+
+    @local_optimizer(None)
+    def local_transform(node):
+        if node in nodes_seen:
+            return False
+        # FIXME: replacing inputs won't work because they are not
+        # outputs of any Apply node
```

Comment:
If you want to support input replacements, you can do it after the optimization with your 'local_transform', loop over the inputs of the optimized graph and get their replacements if applicable and then perform a clone with replacement.